tc/core/polyhedral/cuda/codegen.cc

```diff
@@ -681,6 +704,38 @@ size_t& nAstNodes() {
   return n;
 }
 
+std::unordered_set<isl::id, isl::IslIdIslHash> gatherReadOnlySet(
+    const MappedScop& mscop) {
+  auto root = mscop.schedule();
+  const auto& scop = mscop.scop();
+  std::unordered_set<isl::id, isl::IslIdIslHash> readOnlySet;
+  if (mscop.useReadOnlyCache) {
+    // TensorReferenceGroup::accessedBySubtree seems to require a
+    // MappingFilter. It does not gather references on the root only, so we
+    // iterate on all MappingFilter but ideally scubtrees == root
+    auto subtrees = detail::ScheduleTree::collect(
+        root, detail::ScheduleTreeType::MappingFilter);
+    std::unordered_map<isl::id, bool, isl::IslIdIslHash> isReadOnly;
+    for (auto t : subtrees) {
+      auto groupMap = TensorReferenceGroup::accessedBySubtree(t, scop);
+      for (const auto& kvp : groupMap) {
+        if (isReadOnly.count(kvp.first) == 0) {
+          isReadOnly[kvp.first] = true;
+        }
+        for (const auto& group : kvp.second) {
+          isReadOnly[kvp.first] &= group->isReadOnly();
+        }
+      }
+    }
```

Comment:
This looks like you are abusing a function for something that it was not meant for.
(Not entirely sure because `accessedBySubtree` is not documented.)

Try `scop.reads.universe().range().subtract(scop.writes.universe().range())` instead.